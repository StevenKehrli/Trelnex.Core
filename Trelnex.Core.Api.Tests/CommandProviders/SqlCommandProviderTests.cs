using Azure.Core;
using Azure.Identity;
using Microsoft.Data.SqlClient;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Trelnex.Core.Api.CommandProviders;
using Trelnex.Core.Api.Serilog;
using Trelnex.Core.Data;
using Trelnex.Core.Data.Tests.CommandProviders;

namespace Trelnex.Core.Api.Tests.CommandProviders;

[Ignore("Requires a SQL server.")]
public class SqlCommandProviderTests : CommandProviderTests
{
    private readonly string _scope = "https://database.windows.net/.default";

    private TokenCredential _tokenCredential = null!;
    private string _connectionString = null!;
    private string _tableName = null!;

    [OneTimeSetUp]
    public void TestFixtureSetup()
    {
        // This method is called once prior to executing any of the tests in the fixture.

        // create the service collection
        var services = new ServiceCollection();

        // create the test configuration
        var configuration = new ConfigurationBuilder()
            .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
            .AddJsonFile("appsettings.User.json", optional: true, reloadOnChange: true)
            .Build();

        var dataSource = configuration
            .GetSection("SqlCommandProviders:DataSource")
            .Value;

        var initialCatalog = configuration
            .GetSection("SqlCommandProviders:InitialCatalog")
            .Value;

        _tableName = configuration
            .GetSection("SqlCommandProviders:Tables:0:TableName")
            .Value!;

        var scsBuilder = new SqlConnectionStringBuilder()
        {
            DataSource = dataSource,
            InitialCatalog = initialCatalog,
            Encrypt = true,
        };

        _connectionString = scsBuilder.ConnectionString;

        // create the command provider
        _tokenCredential = new DefaultAzureCredential();

        var bootstrapLogger = services.AddSerilog(
            configuration,
            "Trelnex.Integration.Tests");

        services.AddSqlCommandProviders(
            configuration,
            bootstrapLogger,
            options => options.Add<ITestItem, TestItem>(
                typeName: "test-item",
                validator: TestItem.Validator,
                commandOperations: CommandOperations.All));

        var serviceProvider = services.BuildServiceProvider();

        // get the command provider
        _commandProvider = serviceProvider.GetRequiredService<ICommandProvider<ITestItem>>();
    }

    [TearDown]
    public void Cleanup()
    {
        // This method is called after each test has run.
        using var sqlConnection = new SqlConnection(_connectionString);

        var tokenRequestContext = new TokenRequestContext([ _scope ]);
        sqlConnection.AccessToken = _tokenCredential.GetToken(tokenRequestContext, default).Token;

        sqlConnection.Open();

        var cmdText = $"DELETE FROM [{_tableName}-events]; DELETE FROM [{_tableName}];";
        var sqlCommand = new SqlCommand(cmdText, sqlConnection);

        sqlCommand.ExecuteNonQuery();
    }
}
