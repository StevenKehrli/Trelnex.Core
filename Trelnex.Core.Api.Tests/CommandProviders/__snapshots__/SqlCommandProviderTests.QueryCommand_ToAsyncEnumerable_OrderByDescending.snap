[
  {
    "Item": {
      "Id": "6526e1bc-6344-4091-ae71-3ccf9e2539bc",
      "PartitionKey": "55878311-76a4-42da-8d26-ee047ba7b542",
      "TypeName": "test-item",
      "CreatedDate": "2025-01-04T03:34:09.595945Z",
      "UpdatedDate": "2025-01-04T03:34:09.595945Z",
      "DeletedDate": null,
      "IsDeleted": null,
      "ETag": "fe17a0d9-e693-4350-bcf8-6f6a5dad0afb",
      "Message": "Message #2"
    }
  },
  {
    "Item": {
      "Id": "7429a820-8565-4866-af2d-fd547c7a2935",
      "PartitionKey": "8c420725-2228-4fdb-975e-4545a87a9dce",
      "TypeName": "test-item",
      "CreatedDate": "2025-01-04T03:34:09.595714Z",
      "UpdatedDate": "2025-01-04T03:34:09.595714Z",
      "DeletedDate": null,
      "IsDeleted": null,
      "ETag": "7dfa461e-ac81-4ff3-a985-03d18011a303",
      "Message": "Message #1"
    }
  }
]
